### PR TITLE
KT-31055 Inspection + quickfix to replace mapIndexed with List generator

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-31055

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+            val callExpression = descriptor.psiElement.getStrictParentOfType<KtCallExpression>() ?: return
+            val qualifiedExpression = callExpression.getQualifiedExpressionForSelector() ?: return
+            val receiverExpression = qualifiedExpression.receiverExpression
+            val lambda = callExpression.lambdaArguments.singleOrNull()?.getArgumentExpression() as? KtLambdaExpression ?: return
+
+            val psiFactory = KtPsiFactory(callExpression)
+            val functionLiteral = lambda.functionLiteral
+            functionLiteral.valueParameterList?.replace(psiFactory.createLambdaParameterList(lambda.valueParameters.first().text))
+            functionLiteral.collectDescendantsOfType<KtReturnExpression>().forEach {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt]: Incorrect result:
```kotlin
emptyList<Pair<Int, Int>>().mapIndexed { index, _ ->
    emptyList<String>().mapIndexed { number, s ->
        if (s.isEmpty()) return@mapIndexed number
        else 42
    }
    
    return@mapIndexed index
}
```

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            is KtLambdaExpression -> argumentExpression.valueParameters
+            is KtNamedFunction -> argumentExpression.valueParameters
+            else -> return
+        }
+        if (valueParameters.size != 2) return
+        val secondParameter = valueParameters[1]
+        val destructuringDeclaration = secondParameter.destructuringDeclaration
+        if (destructuringDeclaration != null) {
+            if (destructuringDeclaration.entries.any { entry -> ReferencesSearch.search(entry).any() }) return
+        } else {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt]: I'd prefer to merge `else if`.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        val valueParameters = when (val argumentExpression = valueArgument.getArgumentExpression()) {
+            is KtLambdaExpression -> argumentExpression.valueParameters
+            is KtNamedFunction -> argumentExpression.valueParameters
+            else -> return
+        }
+        if (valueParameters.size != 2) return
+        val secondParameter = valueParameters[1]
+        val destructuringDeclaration = secondParameter.destructuringDeclaration
+        if (destructuringDeclaration != null) {
+            if (destructuringDeclaration.entries.any { entry -> ReferencesSearch.search(entry).any() }) return
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt]: Use `entry.hasUsages(callExpression)` instead.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        } else {
+            if (ReferencesSearch.search(secondParameter).any()) return
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        if (!resolvedCall.isCalling(MAP_INDEXED_FQ_NAME)) return
+        val receiver = callExpression.getQualifiedExpressionForSelector()?.receiverExpression
+        val receiverType = receiver?.getResolvedCall(context)?.resultingDescriptor?.returnType
+            ?: resolvedCall.getImplicitReceiverValue()?.type ?: return
+        if (!receiverType.isCollection(DefaultBuiltIns.Instance)) return
+        val valueArgument = callExpression.valueArguments.singleOrNull() ?: callExpression.lambdaArguments.singleOrNull() ?: return
+        val valueParameters = when (val argumentExpression = valueArgument.getArgumentExpression()) {
+            is KtLambdaExpression -> argumentExpression.valueParameters
+            is KtNamedFunction -> argumentExpression.valueParameters
+            else -> return
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceMapIndexedWithListGeneratorInspection.kt]: False negative:
```kotlin
fun test(list: List<*>) {
    list.mapIndexed label@{ index, _ ->

    }
}
```

**Comment by t-kameyama** [comment]
https://github.com/JetBrains/intellij-community/pull/1615

**Pull request closed without merging**